### ISSUE TITLE
Null pointer dereferences & memory leaks in restart.c

### ISSUE THREAD

**Issue by shouc** [memcached/memcached#595]
### Null pointer dereferencing / Line 33-54
```c
void restart_register(const char *tag, restart_check_cb ccb, restart_save_cb scb, void *data) {
    // when no more available space to allocate, then it would return nullptr, so cb may be nullptr
    restart_data_cb *cb = calloc(1, sizeof(restart_data_cb));
    if (cb_stack == NULL) {
        cb_stack = cb;
    } else {
        restart_data_cb *finder = cb_stack;
        while (finder->next != NULL) {
            finder = finder->next;
        }
        finder->next = cb;
    }
    // cb is dereferenced below
    safe_strcpy(cb->tag, tag, RESTART_TAG_MAXLEN);
    cb->data = data;
    cb->ccb = *ccb;
    cb->scb = *scb;
}
```
### Memory Leak +  Null pointer dereferencing / Line 68-80
```c
static int restart_check(const char *file) {
    size_t flen = strlen(file);
    const char *ext = ".meta";
    // when no more available space to allocate, then it would return nullptr, so metafile may be nullptr
    char *metafile = calloc(1, flen + strlen(ext) + 1);
    // metafile is dereferenced and modified below
    memcpy(metafile, file, flen);
    memcpy(metafile+flen, ext, strlen(ext));

    FILE *f = fopen(metafile, "r");
    if (f == NULL) {
        fprintf(stderr, "[restart] no metadata save file, starting with a clean cache\n");
        // metafile is never freed in this branch
        return -1;
    }

```
### Memory Leak +  Null pointer dereferencing / Line 205-223
```c
static int restart_save(const char *file) {
    size_t flen = strlen(file);
    const char *ext = ".meta";
    // when no more available space to allocate, then it would return nullptr, so metafile may be nullptr
    char *metafile = calloc(1, flen + strlen(ext) + 1);
    // metafile is dereferenced and modified below
    memcpy(metafile+flen, ext, strlen(ext));
    mode_t oldmask = umask(~(S_IRUSR | S_IWUSR));
    FILE *f = fopen(metafile, "w");
    umask(oldmask);
    if (f == NULL) {
        perror("failed to write metadata file");
        return -1;
        // metafile is never freed in this branch
    }
    restart_data_cb *cb = cb_stack;
    restart_cb_ctx ctx;
    ctx.f = f;
    while (cb != NULL) {
        fprintf(f, "T%s\n", cb->tag);
        if (cb->scb(cb->tag, &ctx, cb->data) != 0) {
            // metafile is never freed in this branch
            return -1;
        }

        cb = cb->next;
    }
```

**Comment by dormando** [comment]
Thanks! The leaks were already fixed in the `next` branch. Are the "null pointer dereferencing" you refer to all of the callocs-without-verifications?

**Comment by shouc** [comment]
Yes

**Comment by dormando** [comment]
fixed in 1.5.21, thanks!
